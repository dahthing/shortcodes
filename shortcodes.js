const shortcodes = {};

const add = (generic, callback) => {
  if (typeof generic === 'string' && callback) {
    shortcodes[generic] = callback;
  } else {
    Object.keys(generic).forEach((name) => {
      shortcodes[name] = generic[name];
    });
  }
};

const parse = (input) => {

  return input.replace(/(?=[^\]])\(([a-z0-9_-]+):(.*?)\)/gmi, (match, name, values) => {

    if (!shortcodes[name]) {
      return match;
    }

    const attrs = {};
    const splitted = values.split(/(\s*([a-z0-9_-]+)\:\s*)/gmi);

    attrs[name] = splitted.shift().trim();

    for (let i = 0; i < splitted.length; i += 3) {
      attrs[splitted[i + 1].trim()] =  splitted[i + 2].trim();
    }

    return shortcodes[name](attrs);

  });

};

module.exports = {
  add,
  parse
};
