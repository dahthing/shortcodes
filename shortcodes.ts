export type ShortcodeCallbackAttributes = {
  [key: string]: string;
};

export type ShortcodeCallback = (attrs: ShortcodeCallbackAttributes) => string;

export type ShortcodeMap = {
  [key: string]: ShortcodeCallback;
};

const shortcodes: ShortcodeMap = {};

export const add = (
  generic: string | ShortcodeMap,
  callback?: ShortcodeCallback,
): void => {
  if (typeof generic === 'string' && callback) {
    shortcodes[generic] = callback;
  } else {
    Object.keys(generic).forEach(name => {
      shortcodes[name] = (generic as ShortcodeMap)[name];
    });
  }
};

export const parse = (input: string): string =>
  input.replace(
    /(?=[^\]])\(([a-z0-9_-]+):(.*?)\)/gim,
    (match, name, values) => {
      if (!shortcodes[name]) {
        return match;
      }

      const attrs: ShortcodeCallbackAttributes = {};
      const splitted = values.split(/(\s*([a-z0-9_-]+):\s*)/gim);

      attrs[name] = splitted.shift().trim();

      for (let i = 0; i < splitted.length; i += 3) {
        attrs[splitted[i + 1].trim()] = splitted[i + 2].trim();
      }

      return shortcodes[name](attrs);
    },
  );
