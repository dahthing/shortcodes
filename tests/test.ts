import { readFileSync } from 'fs';
import { resolve } from 'path';
import { add, parse } from '../shortcodes';

// API

describe('API', () => {
  it('provides a proper interface', () => {
    expect(typeof add).toBe('function');
    expect(typeof parse).toBe('function');
  });
});

// Parsing

describe('Parsing', () => {
  add(
    'youtube',
    attrs =>
      `<div class="youtube" style="padding-top:${(100 /
        parseInt(attrs.width, 10)) *
        parseInt(
          attrs.height,
          10,
        )}%"><iframe src="https://www.youtube.com/embed/${
        attrs.youtube
      }" frameborder="0"></iframe></div>`,
  );

  add({
    pow: attrs => String(parseInt(attrs.pow, 10) ** 2),
    sqrt: attrs => String(Math.sqrt(parseInt(attrs.sqrt, 10))),
  });

  ['works', 'parameters', 'ignore'].forEach(test => {
    const input = readFileSync(
      resolve(__dirname, 'fixtures', `${test}.txt`),
      'utf8',
    );

    it(test, () => {
      expect(parse(input)).toMatchSnapshot();
    });
  });
});
