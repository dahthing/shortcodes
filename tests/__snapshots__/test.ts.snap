// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parsing ignore 1`] = `
"Lorem ipsum dolor sit amet, consetetur sadipscing elitr.

(someInvalidShortcodeKey: 123)

Sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat.
"
`;

exports[`Parsing parameters 1`] = `
"Lorem ipsum dolor sit amet, consetetur sadipscing elitr.

<div class=\\"youtube\\" style=\\"padding-top:66.66666666666666%\\"><iframe src=\\"https://www.youtube.com/embed/jNQXAC9IVRw\\" frameborder=\\"0\\"></iframe></div>

Sed diam nonumy eirmod tempor invidunt ut labore et.

<div class=\\"youtube\\" style=\\"padding-top:56.25%\\"><iframe src=\\"https://www.youtube.com/embed/DLzxrzFCyOs\\" frameborder=\\"0\\"></iframe></div>

Dolore magna aliquyam erat, sed diam voluptua.
"
`;

exports[`Parsing works 1`] = `
"Lorem ipsum dolor sit amet, consetetur sadipscing elitr.

15129

Sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat.

3
"
`;
